internal/staging/primitives/subsidy.go

```diff
@@ -0,0 +1,458 @@
+// would be cumbersome to use for callers as it would require them to keep
+// multiple versions of the cache around.  It also would be less efficient
+// because cache reuse would not be possible.
+//
+// Keeping the aforementioned in mind, the preferred way to introduce any such
+// changes is to introduce a new type for the cache, such as SubsidyCacheV2,
+// with the relevant calculation funcs defined on the new type updated to accept
+// the new flags so that the updated version cache can be used for both whatever
+// the new semantics are as well as all historical semantics.
+//
+// Since the cache only stores the full block subsidies prior to any splits or
+// reductions, so long as the consensus change does not involve modifying the
+// full block subsidies, using the same new version of a cache with flags for
+// different splits and/or reduction semantics is straightforward.  On the other
+// hand, supporting a consensus change that involves modifying the full block
+// subsidies requires greater care, but can still be done efficiently by storing
+// additional information regarding which semantics were used to generate the
+// cached entries and reacting accordingly.
+//
+// In the case there are new per-network parameters that must be accepted, a new
+// type for the params, such as SubsidyParamsV2 should be introduced and passed
+// to the constructor of the new cache type (e.g. NewSubsidyCacheV2) given
+// modification of the existing params interface (including adding new methods)
+// would be a major breaking change to the API.  Note that the new version of
+// the parameters will need all of the existing parameters as well, so that can
+// be achieved by embedding the existing params interface in the new one.
+//
+// Finally, the existing type and methods (e.g. SubsidyCache and all calculation
+// methods it provides) that are now handled by the newly introduced versions
+// should then be marked as deprecated with a message instructing the caller to
+// use the newest version of them instead.
+// -----------------------------------------------------------------------------
+
+// SubsidyParams defines an interface that is used to provide the parameters
+// required when calculating block and vote subsidies.  These values are
+// typically well-defined and unique per network.
+type SubsidyParams interface {
+	// BlockOneSubsidy returns the total subsidy of block height 1 for the
+	// network.  This is separate since it encompasses the initial coin
+	// distribution.
+	BlockOneSubsidy() int64
+
+	// BaseSubsidyValue returns the starting base max potential subsidy amount
+	// for mined blocks.  This value is reduced over time and then split
+	// proportionally between PoW, PoS, and the Treasury.  The reduction is
+	// controlled by the SubsidyReductionInterval, SubsidyReductionMultiplier,
+	// and SubsidyReductionDivisor parameters.
+	//
+	// NOTE: BaseSubsidy must be a max of 140,739,635,871,744 atoms or incorrect
```

Comment:
```suggestion
	// NOTE: BaseSubsidyValue must be a max of 140,739,635,871,744 atoms or incorrect
```

Reply:
Good catch.  I just changed it to say "This must be ...".